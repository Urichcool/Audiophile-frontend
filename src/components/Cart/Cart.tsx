import { FC, useEffect } from "react";
import {
  useAppDispatch,
  useAppSelector,
} from "../../redux/reduxHooks/reduxHooks";
import {
  clearCart,
  selectCartProducts,
  selectIsCartModalOpen,
  switchCartModal,
} from "../../redux/slices/cartSlice";
import { BsFillCartXFill } from "react-icons/bs";

const Cart: FC = () => {
  const isCartModalOpen: boolean = useAppSelector(selectIsCartModalOpen);
  const cartProducts:
    | {
        id: string;
        name: string;
        quantity: number;
        price: number;
        picture: string;
      }[]
    | [] = useAppSelector(selectCartProducts);
  const dispatch = useAppDispatch();

  useEffect(() => {
    const handleKeyDown = (e: KeyboardEvent): void => {
      if (e.code === "Escape") {
        dispatch(switchCartModal(false));
      }
    };
    window.addEventListener("keydown", handleKeyDown);
    return (): void => {
      window.removeEventListener("keydown", handleKeyDown);
    };
  }, [dispatch, isCartModalOpen]);

  return (
    <>
      <div
        onClick={(): void => {
          dispatch(switchCartModal(!isCartModalOpen));
        }}
        className={`cart-backdrop cart-backdrop-${
          isCartModalOpen ? "open" : "close"
        }`}
      ></div>
      <div
        className={`cart-container cart-container-${
          isCartModalOpen ? "open" : "close"
        }`}
      >
        <div className="container">
          <div className="cart-content">
            {cartProducts.length === 0 ? (
              <div className="empty-cart-container">
                <h6 className="cart-title">Your cart is empty</h6>
                <BsFillCartXFill />
              </div>
            ) : (
              <div className="cart-head-container">
                <h6 className="cart-title">
                  cart {`(${cartProducts.length})`}
                </h6>
                  <button className="cart-remove-all-button Body-manrope-medium" onClick={(): void => {
                    dispatch(clearCart([]))
                }}>
                  Remove all
                </button>
              </div>
            )}
          </div>
        </div>
      </div>
      ;
    </>
  );
};

export default Cart;
