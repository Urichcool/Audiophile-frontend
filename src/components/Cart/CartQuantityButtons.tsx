import { FC } from "react";
import { AppDispatch } from "../../redux/store";
import { useAppDispatch } from "../../redux/reduxHooks/reduxHooks";
import {
  decreaseQuantity,
  increaseQuantity,
} from "../../redux/slices/cart/cartSlice";

interface ICartQuantityButtons {
  id: string;
  quantity: number;
  price: number;
}

const CartQuantityButtons: FC<ICartQuantityButtons> = ({
  id,
  quantity,
  price,
}) => {
  const dispatch: AppDispatch = useAppDispatch();

  return (
    <div className="cart-quantity-buttons-container">
      <button
        className="cart-quantity-button"
        onClick={(): void => {
          dispatch(decreaseQuantity({ id: id, quantity: 1, price: price }));
        }}
      >
        <p>-</p>
      </button>
      <p className="cart-quantity-text">{quantity}</p>
      <button
        className="cart-quantity-button"
        onClick={(): void => {
          dispatch(increaseQuantity({ id: id, quantity: 1, price: price }));
        }}
      >
        <p>+</p>
      </button>
    </div>
  );
};

export default CartQuantityButtons;
