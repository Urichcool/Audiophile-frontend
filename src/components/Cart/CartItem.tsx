import { FC } from "react";
import { priceWithCommas } from "../../utils/priceWithCommas";
import CartQuantityButtons from "./CartQuantityButtons";

interface ICartItemProps {
  id: string;
  name: string;
  quantity: number;
  price: number;
  picture: string;
  totalPrice: number;
}

const CartItem: FC<ICartItemProps> = ({
  id,
  name,
  quantity,
  price,
  picture,
  totalPrice
}) => {
  return (
    <li className="cart-list-item" data-testid="cart-item">
      <div className="cart-list-item-picture-container">
        <img src={picture} alt={name} className="cart-item-picture" />
      </div>
      <div className="cart-list-item-text-container">
        <h6 className="cart-item-name">{name}</h6>
        <p className="cart-item-price">{priceWithCommas(totalPrice)}</p>
      </div>
      <CartQuantityButtons id={id} quantity={quantity} price={price}/>
    </li>
  );
};

export default CartItem;
