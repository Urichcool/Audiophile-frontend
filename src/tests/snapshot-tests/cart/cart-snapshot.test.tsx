import Cart from "../../../components/Cart/Cart";
import CartItem from "../../../components/Cart/CartItem";
import { renderWithReduxAndRouter } from "../../helpers/RenderWithReduxAndRouter";
import { cartItemTestProps } from "./cart-item-test-props";
import { screen } from "@testing-library/react";

describe("cart-item-snapshot-tests", () => {
  test("cart-item-should-match-snapshot", () => {
    const {
      id,
      name,
      quantity,
      price,
      picture,
      totalPrice,
      category,
    }: {
      id: string;
      name: string;
      quantity: number;
      price: number;
      picture: string;
      totalPrice: number;
      category: string;
    } = cartItemTestProps;
    renderWithReduxAndRouter(
      <CartItem
        id={id}
        name={name}
        quantity={quantity}
        price={price}
        picture={picture}
        totalPrice={totalPrice}
        category={category}
        isStockCheckFetchingHandler={() => {}}
      />
    );
    const cartItem: HTMLLIElement = screen.getByTestId("cart-item");
    expect(cartItem).toMatchSnapshot();
  });
  test("cart-empty-container-should-match-snapshot", () => {
    renderWithReduxAndRouter(<Cart />);
    const cartEmptyContainer: HTMLDivElement = screen.getByTestId(
      "cart-modal-empty-container"
    );
    expect(cartEmptyContainer).toMatchSnapshot();
  });
});
