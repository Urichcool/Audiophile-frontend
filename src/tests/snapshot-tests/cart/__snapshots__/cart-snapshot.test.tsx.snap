// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cart-item-snapshot-tests cart-empty-container-should-match-snapshot 1`] = `
<div
  class="empty-cart-container"
  data-testid="cart-modal-empty-container"
>
  <h6
    class="cart-title"
  >
    Your cart is empty
  </h6>
  <svg
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M.5 1a.5.5 0 0 0 0 1h1.11l.401 1.607 1.498 7.985A.5.5 0 0 0 4 12h1a2 2 0 1 0 0 4 2 2 0 0 0 0-4h7a2 2 0 1 0 0 4 2 2 0 0 0 0-4h1a.5.5 0 0 0 .491-.408l1.5-8A.5.5 0 0 0 14.5 3H2.89l-.405-1.621A.5.5 0 0 0 2 1H.5zM6 14a1 1 0 1 1-2 0 1 1 0 0 1 2 0zm7 0a1 1 0 1 1-2 0 1 1 0 0 1 2 0zM7.354 5.646 8.5 6.793l1.146-1.147a.5.5 0 0 1 .708.708L9.207 7.5l1.147 1.146a.5.5 0 0 1-.708.708L8.5 8.207 7.354 9.354a.5.5 0 1 1-.708-.708L7.793 7.5 6.646 6.354a.5.5 0 1 1 .708-.708z"
    />
  </svg>
</div>
`;

exports[`cart-item-snapshot-tests cart-item-should-match-snapshot 1`] = `
<li
  class="cart-list-item"
  data-testid="cart-item"
>
  <div
    class="cart-list-item-picture-container"
  >
    <img
      alt="YX1 Wireless Earphones"
      class="cart-item-picture"
      src="https://res.cloudinary.com/dir4brsjf/image/upload/v1697496515/audiophile/goods/product-yx1-earphones/desktop/wv1xzajrcrygkt1zarto.jpg"
    />
  </div>
  <div
    class="cart-list-item-text-container"
  >
    <h6
      class="cart-item-name"
    >
      YX1 Wireless Earphones
    </h6>
    <p
      class="cart-item-price"
    >
      $ 1,198
    </p>
  </div>
  <div
    class="cart-quantity-buttons-container"
  >
    <button
      class="cart-quantity-button"
      data-testid="cart-decrease-button"
    >
      <p>
        -
      </p>
    </button>
    <p
      class="cart-quantity-text"
      data-testid="cart-quantity"
    >
      3
    </p>
    <button
      class="cart-quantity-button"
      data-testid="cart-increase-button"
    >
      <p>
        +
      </p>
    </button>
  </div>
</li>
`;
